ext/crates/fp/src/vector/internal/mod.rs

```diff
@@ -0,0 +1,686 @@
+use std::cmp::Ordering;
+
+use super::generic::{FpVectorIterator, FpVectorNonZeroIteratorP, FpVectorP, SliceMutP, SliceP};
+use crate::{
+    constants,
+    limb::{self, Limb, LimbLength},
+    prime::ValidPrime,
+};
+
+mod impl_internal;
+
+pub trait InternalBaseVectorP<const P: u32>: Sized {
+    /// Returns a pointer to the allocation containing the actual data. This is a raw pointer and
+    /// does not take lifetimes into account. It is the responsibility of the caller to ensure that
+    /// the pointer is not dereferenced after the allocation is freed.
+    ///
+    /// We use a pointer instead of a slice because otherwise handling the lifetimes is a huge mess.
+    /// In practice it is almost always better to use [`InternalBaseVectorP::_limbs`] to manipulate
+    /// the underlying data.
+    fn _as_ptr(&self) -> *const Limb;
+
```

Comment:
It seems like this is only used to implement _limbs and _slice, and _slice can be implemented in terms of _limbs. Is there really a need for this function?

Reply:
That kind of mirrors the `InternalBaseMatrixP` trait that does need a raw pointer. I figure already having a raw pointer here might make the interplay easier.

I tried just now to yank `_as_ptr` and make everything depend on `_limbs` but I'm running into a bunch of issues due to lifetimes. There might be a way of handling that gracefully that I'm not seeing, probably because it's getting late

Reply:
What lifetime issues are you running into?

Reply:
So I just made the changes and it looks like `fp` works just fine, but the functionality is slightly different. I was actually running into that same issue before I realized I needed that `Self: 'a` bound. The `algebra::free_module::MuFreeModule::slice_vector` method returns a slice of a slice, and the compiler complains that `v` doesn't live long enough. Having that `Self: 'a` bound makes it behave more like a real `std::slice` slice.

Reply:
I'm not convinced InternalBaseMatrixP requires a raw pointer. In any case I think we should minimize the amount of unsafe needed.

Reply:
(the pointer can be obtained from _limbs anyway)

Reply:
It needs to because in the case of a `MatrixSlice` the limbs won't form a contiguous region in memory if it has fewer columns than the ambient matrix. I also would rather avoid unsafe code but I don't see how we can have the appropriate lifetimes that make `slice_vector` compile.

Reply:
Do we have a need for slicing by columns? I remember this was desired so that we can shard row reduction, but it seems like this is no longer necessary with Nassau's algorithm?

Reply:
(or rather, do we have a need for holding multiple mutable slices in disjoint columns?)

Reply:
I think we definitely need that if we want concurrent algorithms. We probably don't need cutting-edge row reduction algorithms anymore but we can still implement fast matrix-matrix operations. I believe there's a few places where it can be useful, but I'm not familiar enough with the code outside of `fp`. One idea was to use exactly that for #102